### PR TITLE
[TOSA] MultiheadAttention legalization

### PR DESCRIPTION
**Pull request by catcor01**

### CODE REVIEW

File: lib/Conversion/TorchToTosa/TorchToTosa.cpp [hunk 1]
```diff
@@ -9387,6 +9407,32 @@ class ConvertTorchToTosa
 };
 } // namespace
 
+namespace {
+class FoldStaticToDynamicTensorCast
```
Comment (Lallapallooza) [inline lib/Conversion/TorchToTosa/TorchToTosa.cpp]: With `sourceType == resultType` filtered out and differing static shapes rejected by the `tensor.cast` verifier, the new folding pattern never matches. Could we remove this pass?

File: projects/pt1/test/python/scaled_dot_product_attention_lowering.py
Comment (Lallapallooza) [inline projects/pt1/test/python/scaled_dot_product_attention_lowering.py:1]: Given that SDPA is already exercised by the e2e test suite, is a separate standalone script still needed here, or could its coverage be folded into the existing tests instead?

File: lib/Dialect/TorchConversion/Transforms/BackendTypeConversion.cpp [hunk 2]
```diff
@@ -40,6 +41,23 @@ static void setupValueTensorToBuiltinTensorConversion(
       return {};
     return ToBuiltinTensorOp::create(builder, loc, type, inputs[0]);
   });
+  typeConverter.addTargetMaterialization([](OpBuilder &builder, Type type,
```
Comment (Lallapallooza) [inline lib/Dialect/TorchConversion/Transforms/BackendTypeConversion.cpp:44]: The new target materialization creates a `tensor.cast` without checking shape compatibility beyond "result has a static shape". To avoid constructing invalid casts it is safer to use `tensor::CastOp::areCastCompatible`.

File: lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp [hunk 3]
```diff
@@ -2295,6 +2295,218 @@ class DecomposeAtenTraceOp : public OpRewritePattern<AtenTraceOp> {
...
+    auto getOrFallback = [&](ArrayRef<int64_t> staticDims, unsigned idx,
+                             Value fallback) -> Value {
+      return getDimValue(idx < staticDims.size() ? staticDims[idx]
+                                                 : Torch::kUnknownSize,
+                         fallback);
+    };
+    keyDims.push_back(getOrFallback(keyTransposedSizes, 0, batchSize));
+    if (keyTransposedSizes.size() == 4) {
+      keyDims.push_back(getOrFallback(keyTransposedSizes, 1, numHeadsSize));
+      keyDims.push_back(getOrFallback(keyTransposedSizes, 2, seqLen));
```
Comment (Lallapallooza) [inline lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp]: After the transpose the third axis now holds the head dimension, not the query sequence length. Using seqLen as the fallback here couples the head axis to the sequence length for dynamic cases and is incorrect when seqLen and headDim differ.

File: lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp [hunk 4]
```diff
@@ -2295,6 +2295,218 @@ class DecomposeAtenTraceOp : public OpRewritePattern<AtenTraceOp> {
...
+    if (!queryTensorType || !keyTensorType || !valueTensorType)
+      return rewriter.notifyMatchFailure(op, "expected tensor inputs");
+    if (!queryTensorType.hasSizes() || !keyTensorType.hasSizes() ||
+        !valueTensorType.hasSizes())
+      return rewriter.notifyMatchFailure(
+          op, "expected tensor inputs to have known shapes");
+    auto queryValueTensorType = dyn_cast<ValueTensorType>(queryTensorType);
+    auto keyValueTensorType = dyn_cast<ValueTensorType>(keyTensorType);
+    auto valueValueTensorType = dyn_cast<ValueTensorType>(valueTensorType);
+    if (!queryValueTensorType || !keyValueTensorType || !valueValueTensorType)
```
Comment (Lallapallooza) [inline lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp:2342]: The decomposition assumes compatible dtypes for query/key/value but never checks them. Verifying that QKV share the same `getOptionalDtype()` would catch inconsistent inputs early.

File: lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp [hunk 5]
```diff
@@ -2295,6 +2295,218 @@ class DecomposeAtenTraceOp : public OpRewritePattern<AtenTraceOp> {
...
+    Value keepDimTrue =
+        ConstantBoolOp::create(rewriter, loc, rewriter.getBoolAttr(true));
+    auto maxOp =
+        AtenMaxDimOp::create(rewriter, loc, reducedValueType, reducedIndexType,
+                             scaledScores, negOne, keepDimTrue);
+    Value softmaxMax = TensorStaticInfoCastOp::create(
+        rewriter, loc, reducedValueType, maxOp.getValues());
+    Value centered =
+        createTensorSub(rewriter, loc, scoresType, scaledScores, softmaxMax);
+    Value unNormalizedExp =
```
Comment (Lallapallooza) [inline lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp]: This reimplements the max‑shifted softmax in the input tensor's dtype, whereas `getSoftmaxResult` just below already centralizes this logic.

File: lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp [hunk 6]
```diff
@@ -2295,6 +2295,218 @@ class DecomposeAtenTraceOp : public OpRewritePattern<AtenTraceOp> {
...
+    Value zeroInt =
+        ConstantIntOp::create(rewriter, loc, rewriter.getI64IntegerAttr(0));
+    Value rank = AtenDimOp::create(rewriter, loc, query);
+    Value lastDim = AtenSubIntOp::create(rewriter, loc, rank, oneInt);
+    Value headDim = AtenSizeIntOp::create(rewriter, loc, query, lastDim);
+    Value seqDimIndex = AtenSubIntOp::create(rewriter, loc, lastDim, oneInt);
+    Value seqLen = AtenSizeIntOp::create(rewriter, loc, query, seqDimIndex);
+    Value keySeqLen = AtenSizeIntOp::create(rewriter, loc, key, seqDimIndex);
+    ArrayRef<int64_t> querySizes = queryValueTensorType.getSizes();
+    bool hasExplicitHeadDim = querySizes.size() >= 4;
```
Comment (Lallapallooza) [inline lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp]: This treats every rank >4 tensor as [batch, heads, seq, dim], implicitly dropping the possibility of extra broadcastable prefix dims that the upstream shape function allows, either constraining to rank 3/4 or handling additional leading axes explicitly would make the supported layout contract clearer.

**Review by Lallapallooza — COMMENTED** [review]
Thanks for patch, few comments.

File: projects/pt1/test/python/scaled_dot_product_attention_lowering.py
Comment (catcor01) [reply projects/pt1/test/python/scaled_dot_product_attention_lowering.py:1]: ack yes it is redundant

File: lib/Dialect/TorchConversion/Transforms/BackendTypeConversion.cpp (hunk 2 above)
Comment (catcor01) [reply lib/Dialect/TorchConversion/Transforms/BackendTypeConversion.cpp:44]: ack. Now also check tensor::CastOp::areCastCompatible.

File: lib/Conversion/TorchToTosa/TorchToTosa.cpp (hunk 1 above)
Comment (catcor01) [reply lib/Conversion/TorchToTosa/TorchToTosa.cpp]: ack removed.

File: lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp (hunk 5 above)
Comment (catcor01) [reply lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp]: ack. Refactored getSoftmaxResult to take an explicit dim operand so it can be shared by SDPA and existing softmax decompositions

File: lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp (hunk 6 above)
Comment (catcor01) [reply lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp]: ack.  limit support to the documented 3D/4D layouts

File: lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp (hunk 4 above)
Comment (catcor01) [reply lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp:2342]: ack. done.

File: lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp (hunk 3 above)
Comment (catcor01) [reply lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp]: ack.  fixeed the key-transpose reshape so the third axis tracks the head dimension

File: lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp [hunk 7]
```diff
@@ -2295,17 +2295,223 @@ class DecomposeAtenTraceOp : public OpRewritePattern<AtenTraceOp> {
 };
 } // namespace
 
+static Value getSoftmaxResult(Operation *op, Value self, Value dim,
+                              Type resultType, Type accumulatorType,
+                              PatternRewriter &rewriter);
+
+namespace {
+// Decompose scaled dot product attention into matmul/softmax pipeline when
```
Comment (sahas3) [inline lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp:2303]: Is this decomposition producing any different IR compared to leveraging the decomposition of `sdpa` with `ExportedProgram.run_decompositions` https://docs.pytorch.org/docs/stable/export.html#export-ir-decompositions -- see https://discord.com/channels/636084430946959380/742573221882364009/1446121930922004623 for reference.

I am wondering if the `sdpa` op should be added to the default decomposition list in https://github.com/llvm/torch-mlir/blob/42c3c29d449f3dd0c86f781ff8710a398011ea8c/python/torch_mlir/extras/fx_decomp_util.py#L5 instead?

Comment (catcor01) [reply lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp:2303]: yes you are completely correct, thanks for pointing this out. added.

File: lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp [hunk 8]
```diff
@@ -13084,6 +13292,8 @@ class DecomposeComplexOpsPass
     legalOpsSet.clear();
     legalOpsSet.insert(legalOps.begin(), legalOps.end());
 
+    patterns.add<DecomposeAtenScaledDotProductAttentionOp>(context);
```
Comment (sahas3) [inline lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp]: Is this pattern needed anymore with the change in `fx_decomp_util`?

Comment (catcor01) [reply lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp]: Please correct me if I misunderstand, but I believe we still need the MLIR-side pattern. The new entry in python/torch_mlir/extras/fx_decomp_util.py only affects the FX/ExportedProgram import path. Other frontends—TorchScript, AOTAutograd, or anyone who feeds raw Torch dialect into torch-mlir-opt—never touch that Python list, so they can still produce torch.aten.scaled_dot_product_attention. For those cases the rewrite in lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp is what lowers sdpa into the matmul/softmax pipeline so that downstream -convert-torch-to-tosa or -convert-torch-to-linalg keeps working.

Comment (sahas3) [reply lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp]: Based on https://github.com/llvm/torch-mlir/blob/0844d4d4221f7f9bcc989dbbf6fc1bcde7701a5b/docs/development.md?plain=1#L244 IIUC, `fx_importer` path is the only maintained path. Rest have been deprecated but the code still exists. Maybe @sjarus / @zjgarvey can confirm / correct that understanding and we can discuss if it's still valuable to have this decomposition pattern or we can rely on PyTorch's decomposition. 

I don't have any flag on adding this, just want to make sure that it will actually be exercised.

Comment (zjgarvey) [reply lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp]: Yes, the `fx_importer` is the only path we should expect to support.

I have found attention to be a bit frustrating, however. For example, running decompositions on an exported program with an sdpa op sometimes converts sdpa into a slightly different attention op- even when attention itself isn't getting decomposed. Merely running decompositions at all actually retraces the graph with a different tool, and may select different ops further varied  based on other factors like the torch device used by the inputs. 

In any case, I don't mind adding a decomposition pattern. We have a bit more control with a pattern like this as opposed to fx decompositions.

Comment (zjgarvey) [reply lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp]: Yes, the `fx_importer` is the only path we should expect to support.

I have found attention to be a bit frustrating, however. For example, running decompositions on an exported program with an sdpa op sometimes converts sdpa into a slightly different attention op- even when attention itself isn't getting decomposed. Merely running decompositions at all actually retraces the graph with a different tool, and may select different ops further varied  based on other factors like the torch device used by the inputs. 

In any case, I don't mind adding a decomposition pattern. We have a bit more control with a pattern like this as opposed to fx decompositions.

Comment (catcor01) [reply lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp]: @sahas3 based on the above comment are you happy to keep the decomposition as is or would you prefer to remove?

File: lib/Conversion/TorchToTosa/TorchToTosa.cpp [hunk 9]
```diff
@@ -4016,8 +4016,28 @@ LogicalResult ConvertAtenOp<AtenTransposeIntOp>::matchAndRewrite(
   transposedDims[dim0] = dim1;
   transposedDims[dim1] = dim0;
 
+  Type resultType = getTypeConverter()->convertType(op.getType());
+  if (auto rankedSelf = dyn_cast<RankedTensorType>(selfType)) {
+    SmallVector<int64_t> transposedShape(rankedSelf.getRank(),
+                                         ShapedType::kDynamic);
+    if (rankedSelf.hasStaticShape()) {
+      auto staticShape =
```
Comment (sahas3) [inline lib/Conversion/TorchToTosa/TorchToTosa.cpp]: `to_vector` is redundant since the return type of `makeShapeTorchCompatible` is already `SmallVector`

File: lib/Conversion/TorchToTosa/TorchToTosa.cpp [hunk 10]
```diff
@@ -4016,8 +4016,28 @@ LogicalResult ConvertAtenOp<AtenTransposeIntOp>::matchAndRewrite(
...
+          llvm::to_vector(makeShapeTorchCompatible(rankedSelf.getShape()));
+      auto dim0Index = static_cast<size_t>(dim0);
+      auto dim1Index = static_cast<size_t>(dim1);
+      if (dim0Index < staticShape.size() && dim1Index < staticShape.size())
```
Comment (sahas3) [inline lib/Conversion/TorchToTosa/TorchToTosa.cpp]: Isn't this condition always guaranteed by the check on line https://github.com/llvm/torch-mlir/blob/9ef3ed1e44e6a0288bf8beefc27befd94d4448ab/lib/Conversion/TorchToTosa/TorchToTosa.cpp#L4008 ?

File: lib/Conversion/TorchToTosa/TorchToTosa.cpp [hunk 11]
```diff
@@ -4016,8 +4016,28 @@ LogicalResult ConvertAtenOp<AtenTransposeIntOp>::matchAndRewrite(
...
+        std::swap(staticShape[dim0Index], staticShape[dim1Index]);
+      for (size_t i = 0; i < staticShape.size(); ++i)
+        transposedShape[i] = staticShape[i];
+    }
+    auto rankedResult = RankedTensorType::get(
```
Comment (sahas3) [inline lib/Conversion/TorchToTosa/TorchToTosa.cpp]: IIUC, you are computing the transposed shape for statically shaped inputs and using that to construct `tosa::transposeOp`. I think we can use `tosa::CreateOpAndInfer` with `UnrankedTensorType::get(elemTy)` and let the transposed op creation process infer the resultType instead of computing it here.

File: lib/Conversion/TorchToTosa/TorchToTosa.cpp [hunk 12]
```diff
@@ -9402,6 +9422,21 @@ void populateTorchToTosaConversionLegalOps(ConversionTarget &target) {
...
+    if (sourceType.getElementType() != resultType.getElementType())
+      return true;
+    if (!sourceType.hasStaticShape())
+      return true;
+    if (!resultType.hasStaticShape())
+      return true;
+    if (sourceType == resultType)
+      return true;
+    return false;
+  });
```
Comment (sahas3) [inline lib/Conversion/TorchToTosa/TorchToTosa.cpp:10090]: I am guessing this is needed because of the targetMaterialization change?

File: lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp (hunk 8 above)
Comment (sahas3) [inline lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp]:
```suggestion
    addPatternIfTargetOpIsIllegal<DecomposeAtenScaledDotProductAttentionOp>(patterns);
```

File: lib/Dialect/TorchConversion/Transforms/BackendTypeConversion.cpp [hunk 13]
```diff
@@ -40,6 +41,25 @@ static void setupValueTensorToBuiltinTensorConversion(
...
```
Comment (sahas3) [inline lib/Dialect/TorchConversion/Transforms/BackendTypeConversion.cpp:44]: Clarification: what motivated this change? 

Also, I am not seeing any `tensor.cast` op in the LIT you've added. Is a testcase missing from this PR?

File: test/Conversion/TorchToTosa/multi_head_attention.mlir [hunk 14]
```diff
@@ -0,0 +1,29 @@
+// RUN: torch-mlir-opt %s -torch-decompose-complex-ops -convert-torch-to-tosa -split-input-file | FileCheck %s
```
Comment (sahas3) [inline test/Conversion/TorchToTosa/multi_head_attention.mlir]: Can you also add LIT tests for the negative cases to ensure decomposition doesn't happen?

File: lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp [hunk 15]
```diff
@@ -2295,17 +2295,223 @@ class DecomposeAtenTraceOp : public OpRewritePattern<AtenTraceOp> {
...
+    auto queryValueTensorType = dyn_cast<ValueTensorType>(queryTensorType);
+    auto keyValueTensorType = dyn_cast<ValueTensorType>(keyTensorType);
+    auto valueValueTensorType = dyn_cast<ValueTensorType>(valueTensorType);
+    if (!queryValueTensorType || !keyValueTensorType || !valueValueTensorType)
+      return rewriter.notifyMatchFailure(op, "expected value tensor semantics");
+    if (!queryValueTensorType.hasDtype() || !keyValueTensorType.hasDtype() ||
+        !valueValueTensorType.hasDtype())
+      return rewriter.notifyMatchFailure(
+          op, "expected tensor inputs to have dtypes");
+    Type queryDtype = queryValueTensorType.getOptionalDtype();
```
Comment (sahas3) [inline lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp]:
```suggestion
    Type queryDtype = queryValueTensorType.getDtype();
```
since `hasDtype()` is already successful.

File: lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp [hunk 16]
```diff
@@ -2295,17 +2295,223 @@ class DecomposeAtenTraceOp : public OpRewritePattern<AtenTraceOp> {
...
+    int64_t queryRank = querySizes.size();
+    if (queryRank < 3 || queryRank > 4)
+      return rewriter.notifyMatchFailure(
+          op, "expected query tensor rank to be 3 or 4");
+    ArrayRef<int64_t> keySizes = keyValueTensorType.getSizes();
+    ArrayRef<int64_t> valueSizes = valueValueTensorType.getSizes();
+    if (static_cast<int64_t>(keySizes.size()) != queryRank ||
+        static_cast<int64_t>(valueSizes.size()) != queryRank)
+      return rewriter.notifyMatchFailure(
+          op, "expected query, key, and value to share rank");
```
Comment (sahas3) [inline lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp:2369]: This failure check should happen before any IR modification starts otherwise IR will be in a bad state as we've already introduced new ops but the original op is not replaced, eventually leading to the pass failing.

File: lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp [hunk 17]
```diff
@@ -2295,17 +2295,223 @@ class DecomposeAtenTraceOp : public OpRewritePattern<AtenTraceOp> {
...
+
+    Value negTwo =
+        ConstantIntOp::create(rewriter, loc, rewriter.getI64IntegerAttr(-2));
+    Value negOne =
+        ConstantIntOp::create(rewriter, loc, rewriter.getI64IntegerAttr(-1));
+
+    SmallVector<int64_t> keyTransposedSizes(keySizes.begin(), keySizes.end());
+    if (keyTransposedSizes.size() < 2)
+      return rewriter.notifyMatchFailure(
+          op, "expected key tensor rank >= 2 for transpose");
```
Comment (sahas3) [inline lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp]: this too has to be checker earlier

File: lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp [hunk 18]
```diff
@@ -2295,17 +2295,223 @@ class DecomposeAtenTraceOp : public OpRewritePattern<AtenTraceOp> {
...
+    scores =
+        AtenViewOp::create(rewriter, loc, scoresType, scores, scoresShapeList);
+    Value scaledScores =
+        AtenMulScalarOp::create(rewriter, loc, scoresType, scores, scaleFloat);
+
+    Value softmax = getSoftmaxResult(op.getOperation(), scaledScores, negOne,
+                                     scoresType, scoresType, rewriter);
+    if (!softmax)
+      return rewriter.notifyMatchFailure(op,
+                                         "failed to compute softmax scores");
```
Comment (sahas3) [inline lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp:2487]: This one probably is fine. Looking at `getSoftmaxResult` I don't expect it to fail in this logic. I can't think of a good idea to ensure that though barring replicating the checks in `getSoftmaxResult` here early enough but that doesn't seem like a good idea.

File: test/Conversion/TorchToTosa/multi_head_attention.mlir [hunk 19]
```diff
@@ -0,0 +1,29 @@
...
+
+// Checks that scaled dot product attention (single-head configuration) lowers
```
Comment (sahas3) [inline test/Conversion/TorchToTosa/multi_head_attention.mlir]: I think a better place for this test is https://github.com/llvm/torch-mlir/blob/ac7b5f5d0feb9c07d56ec6a19cb66483b0780f53/test/Dialect/Torch/decompose-complex-ops.mlir#L1 since the decomposition path is common across all backend paths.

File: lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp (hunk 8 above)
Comment (sahas3) [reply lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp]: Thanks for the reminder, I'm fine with adding the pattern.

The puzzle now is that if we add `sdpa` to the decomp list in `fx_decomp_util`, while we do have a LIT test locking down the decomposition pattern, it won't be tested via the `e2e` test. However the ops we are decomposing to with the C++ pattern are all locked down via `e2e` test (I hope :D ). So not being able to `e2e` test this C++ decomp pattern is probably fine? Thoughts @zjgarvey / @sjarus ?

On that note, we probably need a way to be able to control what ops to decompose with PyTorch decomposition at the `e2e` test level but that's out of scope of this PR.

File: lib/Conversion/TorchToTosa/TorchToTosa.cpp (hunk 10 above)
Comment (catcor01) [reply lib/Conversion/TorchToTosa/TorchToTosa.cpp]: ack.

File: lib/Conversion/TorchToTosa/TorchToTosa.cpp (hunk 11 above)
Comment (catcor01) [reply lib/Conversion/TorchToTosa/TorchToTosa.cpp]: ack.

File: lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp (hunk 8 above)
Comment (catcor01) [reply lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp]: done.

File: lib/Conversion/TorchToTosa/TorchToTosa.cpp (hunk 12 above)
Comment (catcor01) [reply lib/Conversion/TorchToTosa/TorchToTosa.cpp:10090]: yes exactly.

File: lib/Dialect/TorchConversion/Transforms/BackendTypeConversion.cpp (hunk 13 above)
Comment (catcor01) [reply lib/Dialect/TorchConversion/Transforms/BackendTypeConversion.cpp:44]: We added that target materialization so TOSA conversions don’t fail when MLIR infers a slightly more precise tensor type than the one produced by the Torch type converter. For example, the flatten test in basic.mlir (line 683) expects a result of type tensor<10x3x?x4xf32>, because the middle dimension is intentionally left dynamic on the Torch side. After the reshape, TOSA inference yields tensor<10x3x216x4xf32> since it can fold all dims statically. The materializer inserts a tensor.cast to reconcile those two types, and the dynamic-legality hook simply says “these identity-like casts are fine to keep.” That existing flatten test already checks for the emitted tensor.cast, so we didn’t need to add another lit file just for this change.

File: test/Conversion/TorchToTosa/multi_head_attention.mlir (hunk 14 above)
Comment (catcor01) [reply test/Conversion/TorchToTosa/multi_head_attention.mlir]: Done. decompose-complex-ops.mlir now contains both the positive SDPA decomposition check and the two negative “keep as-is” cases, and multi_head_attention.mlir has been deleted accordingly.

File: lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp (hunk 16 above)
Comment (catcor01) [reply lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp:2369]: Done.

File: lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp (hunk 17 above)
Comment (catcor01) [reply lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp]: Done.

File: test/Conversion/TorchToTosa/multi_head_attention.mlir (hunk 19 above)
Comment (catcor01) [reply test/Conversion/TorchToTosa/multi_head_attention.mlir]: Done

File: lib/Conversion/TorchToTosa/TorchToTosa.cpp (hunk 9 above)
Comment (catcor01) [reply lib/Conversion/TorchToTosa/TorchToTosa.cpp]: Done.

**Comment by catcor01** [comment]
@sahas3 My apologies for amending the commit again in the previous push, I had forgotten about not doing so.

File: lib/Conversion/TorchToTosa/TorchToTosa.cpp [hunk 20]
```diff
@@ -4474,9 +4474,27 @@ LogicalResult ConvertAtenOp<AtenTransposeIntOp>::matchAndRewrite(
...
+        op, "failed to convert transpose result type");
+
+  auto elementType = cast<TensorType>(selfType).getElementType();
+  auto unrankedResultType = UnrankedTensorType::get(elementType);
+  auto transpose = tosa::CreateOpAndInfer<tosa::TransposeOp>(
+      rewriter, op->getLoc(), unrankedResultType, adaptor.getSelf(),
       rewriter.getDenseI32ArrayAttr(transposedDims));
+  Value resultValue = transpose.getResult();
+  if (resultValue.getType() != expectedResultType) {
+    if (!tensor::CastOp::areCastCompatible(resultValue.getType(),
```
Comment (sahas3) [inline lib/Conversion/TorchToTosa/TorchToTosa.cpp:4489]: Do you expect this scenario to be hit? If it does we are again leaving the IR in a bad state since `transpose` will be inserted but the `torch` op won't be replaced.

File: lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp [hunk 21]
```diff
@@ -2295,17 +2295,221 @@ class DecomposeAtenTraceOp : public OpRewritePattern<AtenTraceOp> {
...
+
+    Value query = op.getQuery();
+    Value key = op.getKey();
+    Value value = op.getValue();
+
+    auto queryTensorType = dyn_cast<BaseTensorType>(query.getType());
+    auto keyTensorType = dyn_cast<BaseTensorType>(key.getType());
+    auto valueTensorType = dyn_cast<BaseTensorType>(value.getType());
+    if (!queryTensorType || !keyTensorType || !valueTensorType)
+      return rewriter.notifyMatchFailure(op, "expected tensor inputs");
```
Comment (sahas3) [inline lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp]: Are these checks required or can we rely on the checks for `ValueTensorType` on 2351 only?

File: lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp [hunk 22]
```diff
@@ -2295,17 +2295,221 @@ class DecomposeAtenTraceOp : public OpRewritePattern<AtenTraceOp> {
...
+    int64_t queryRank = querySizes.size();
+    if (queryRank < 3 || queryRank > 4)
+      return rewriter.notifyMatchFailure(
+          op, "expected query tensor rank to be 3 or 4");
+    ArrayRef<int64_t> keySizes = keyValueTensorType.getSizes();
+    ArrayRef<int64_t> valueSizes = valueValueTensorType.getSizes();
+    if (static_cast<int64_t>(keySizes.size()) != queryRank ||
+        static_cast<int64_t>(valueSizes.size()) != queryRank)
+      return rewriter.notifyMatchFailure(
+          op, "expected query, key, and value to share rank");
```
Comment (sahas3) [inline lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp:2369]: This should happen before 2364 to avoid introducing new IR in case of match failure.

File: lib/Conversion/TorchToTosa/TorchToTosa.cpp (hunk 20 above)
Comment (catcor01) [reply lib/Conversion/TorchToTosa/TorchToTosa.cpp:4489]: Yes, I believe we expect the case and handle it; the true failure case shouldn’t happen, but if it did, the rewriter would undo the partial rewrite and preserve valid IR.

More info:

The branch is only taken when TOSA infers an unranked result but the Torch type converter produced a ranked type. It can happen for inputs that enter the pattern as unranked, and we immediately insert a tensor.cast to reconcile the types, which is intentional.
If areCastCompatible were to fail, the pattern returns failure(). Because ConversionPatternRewriter is transactional, all ops created while the pattern runs (including the temporary transpose) are rolled back, so the IR isn’t left half-transformed.

File: lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp (hunk 21 above)
Comment (catcor01) [reply lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp]: We can rely solely on the ValueTensorType checks at 2351. Every ValueTensorType is already a BaseTensorType, so the earlier dyn_cast<BaseTensorType> + hasSizes() checks were redundant. After the ValueTensorType dyn_cast succeeds we still verify hasSizes()/hasDtype(), so we don’t lose any coverage. I dropped the extra casts in the updated patch.

File: lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp (hunk 22 above)
Comment (catcor01) [reply lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp:2369]: Done.

**Review by Lallapallooza — APPROVED** [review]
